mipengine/controller/cleaner.py

```diff
@@ -0,0 +1,211 @@
+        parsed_toml[context_id]["timestamp"] = now_timestamp.isoformat()
+        parsed_toml[context_id]["released"] = False
+
+        tmp_contextids_cleanup_file = os.path.join(dirname, filename_tmp)
+        with open(tmp_contextids_cleanup_file, "w") as f:
+            toml.dump(parsed_toml, f)
+
+        # renaming is atomic. Of course if more Controllers are spawn the file is very
+        # likely to become corrupted
+        os.rename(
+            tmp_contextids_cleanup_file,
+            controller_config.cleanup.contextids_cleanup_file,
+        )
+
+    def _set_released_true_to_file(self, context_id: str):
+        dirname = os.path.dirname(controller_config.cleanup.contextids_cleanup_file)
+        filename = os.path.basename(controller_config.cleanup.contextids_cleanup_file)
+        filename_tmp = "contextids_cleanup_tmp.toml"
+        with open(controller_config.cleanup.contextids_cleanup_file, "r") as f:
+            parsed_toml = toml.load(f)
+        if context_id in parsed_toml:
+            parsed_toml[context_id]["released"] = True
+        tmp_contextids_cleanup_file = os.path.join(dirname, filename_tmp)
+        with open(tmp_contextids_cleanup_file, "w") as f:
+            toml.dump(parsed_toml, f)
+
+        # renaming is atomic. Of course if more Controllers are spawn the file is very
+        # likely to become corrupted
+        os.rename(
+            tmp_contextids_cleanup_file,
+            controller_config.cleanup.contextids_cleanup_file,
+        )
+
+    def _remove_from_cleanup_file(self, context_id: str, node_id: str = None):
+        dirname = os.path.dirname(controller_config.cleanup.contextids_cleanup_file)
+        filename = os.path.basename(controller_config.cleanup.contextids_cleanup_file)
+        filename_tmp = "contextids_cleanup_tmp.toml"
+        with open(controller_config.cleanup.contextids_cleanup_file, "r") as f:
+            parsed_toml = toml.load(f)
+        if context_id in parsed_toml:
+            if node_id:
+                try:
+                    parsed_toml[context_id]["nodes"].remove(node_id)
+                except ValueError:
+                    # warning if nodeid is not there
+                    pass
+            else:
+                parsed_toml.pop(context_id)
+        else:
+            # warning if contextid not there
+            pass
+
+        tmp_contextids_cleanup_file = os.path.join(dirname, filename_tmp)
+        with open(tmp_contextids_cleanup_file, "w") as f:
+            toml.dump(parsed_toml, f)
+
+        # renaming is atomic. Of course if more Controllers are spawn the file is very
+        # likely to become corrupted
+        os.rename(
+            tmp_contextids_cleanup_file,
+            controller_config.cleanup.contextids_cleanup_file,
+        )
+
+    def _read_cleanup_file(self) -> dict:
+        with open(controller_config.cleanup.contextids_cleanup_file, "r") as f:
+            parsed_toml = toml.load(f)
```

Comment:
check file exists, [commit](https://github.com/madgik/MIP-Engine/pull/205/commits/81cfcbd9a68ee6cc091a86d75fba5e1a281042b1)

Reply:
I think it's safer to have a try catch around the 
```
        with open(controller_config.cleanup.contextids_cleanup_file, "r") as f:
            parsed_toml = toml.load(f)
```

so if something happens that we haven't though about it, it won't crash the cleaner. If something like that happens, the file should be created again, clean, so that the cleaner can continue working. With the current implementation we will just log it, we won't recover from it.

Reply:
The file gets created for the first time (if it does not exist) in \_\_init\_\_ line 125. Then it gets  recreated/overwritten again and again every time `_write_to_cleanup_file` method is called.

Reply:
What if the file got corrupted for some reason and cannot be read? The `toml.load` will fail.

I understand that it's a slim chance if there is a chance at all, but I think it's much safer to wrap your code around a try catch and then recover from the failure, instead of trying to protect from that scenario before running your code. Something that we haven't though might happen and then the code will break without being able to recover.

Reply:
ok yes `toml.load` could raise an exception in that scenario, I ll wrap it in a try and just log it, it should not be recreated there.

Reply:
[commit](https://github.com/madgik/MIP-Engine/pull/205/commits/47c7f1ebd52ccf1f691c57cc09dd8884e8394f9e#diff-6cd1874ceb56337f09b5846eadef72e80f879e9668be711d0b4386d72e6a8686)